apps/web/src/hooks/useDocumentParameters/index.ts

```diff
@@ -1,139 +1,292 @@
+      setInput(source, value, param)
+    },
+    [setInput, source],
+  )
+
+  const setSource = useCallback(
+    (source: InputSource) => {
+      setValue((prev) => {
+        const { state, doc } = getDocState(prev, key)
+        return {
+          ...state,
+          [key]: {
+            ...doc,
+            source,
+          },
+        }
       })
+    },
+    [key, setValue],
+  )
 
-      setInputs(newInputs)
+  const setDataset = useCallback(
+    (selected: DatasetSource) => {
+      setValue((old) => {
+        const { state, doc } = getDocState(old, key)
+        return {
+          ...state,
+          [key]: {
+            ...doc,
+            dataset: {
+              ...doc.dataset,
+              ...selected,
+            },
+          },
+        }
+      })
     },
-    [inputs, newParams, setInputs],
+    [allInputs, key, setValue],
   )
 
-  return { parameters, inputs, setInputs, onMetadataProcessed }
+  const copyDatasetInputsToManual = useCallback(() => {
+    setManualInputs(inputs['dataset'].inputs)
+  }, [inputs])
+
+  const setHistoryLog = useCallback(
+    (logUuid: string) => {
+      setValue((old) => {
+        const { state, doc } = getDocState(old, key)
+        return {
+          ...state,
+          [key]: {
+            ...doc,
+            history: {
+              ...doc.history,
+              logUuid,
+            },
+          },
+        }
+      })
+    },
+    [allInputs, key, setValue],
+  )
+
+  const onMetadataProcessed = useCallback(
+    (metadataParameters: Set<string>) => {
+      setInputs(
+        'manual',
+        recalculateInputs({ inputs: inputs.manual.inputs, metadataParameters }),
+      )
+      setInputs(
+        'dataset',
+        recalculateInputs({
+          inputs: inputs.dataset.inputs,
+          metadataParameters,
+        }),
+      )
+      setInputs(
+        'history',
+        recalculateInputs({
+          inputs: inputs.history.inputs,
+          metadataParameters,
+        }),
+      )
+    },
+    [inputs, setInputs, source],
+  )
+
+  const parameters = useMemo(
+    () => convertToParams(inputsBySource),
+    [inputsBySource],
+  )
+
+  return {
+    parameters,
+    onMetadataProcessed,
+    source,
+    setSource,
+    setInput,
+    manual: {
+      inputs: inputs['manual'].inputs,
+      setInput: setManualInput,
+      setInputs: setManualInputs,
+    },
+    dataset: {
+      datasetId: inputs['dataset'].datasetId,
+      rowIndex: inputs['dataset'].rowIndex,
+      inputs: inputs['dataset'].inputs,
+      mappedInputs: inputs['dataset'].mappedInputs,
+      setInput: setDatasetInput,
+      setInputs: setDatasetInputs,
+      copyToManual: copyDatasetInputsToManual,
+      setDataset,
+    },
+    history: {
+      logUuid: inputs['history'].logUuid,
+      inputs: inputs['history'].inputs,
+      setInput: setHistoryInput,
+      setInputs: setHistoryInputs,
+      setHistoryLog,
+    },
+  }
```

Comment:
This hook is huge and it could have some tests. But I don't want to invest more time here. Also I feel being fully typed is easy to follow. Also all the methods are small